source/common/compressor/zlib_compressor_impl.cc

```diff
@@ -0,0 +1,108 @@
+#include "common/compressor/zlib_compressor_impl.h"
+
+namespace Envoy {
+namespace Compressor {
+
+ZlibCompressorImpl::ZlibCompressorImpl() : zlib_ptr_(new z_stream()) {}
+
+ZlibCompressorImpl::~ZlibCompressorImpl() {
+  // deflateEnd/inflateEnd must be called, otherwise it will cause memory leaks.
+  if (is_deflate_) {
+    deflateEnd(zlib_ptr_.get());
+  } else {
+    inflateEnd(zlib_ptr_.get());
+  }
+}
+
+void ZlibCompressorImpl::setChunk(uint64_t chunk) { chunk_ = chunk; }
+
+bool ZlibCompressorImpl::init(CompressionLevel comp_level, CompressionStrategy comp_strategy,
+                              int window_bits, uint memory_level) {
+  // prevents deflateInit2() of being called multiple times, which would cause memory leaks.
+  if (is_initialized_) {
+    return true;
+  }
+
+  zlib_ptr_->zalloc = Z_NULL;
+  zlib_ptr_->zfree = Z_NULL;
+  zlib_ptr_->opaque = Z_NULL;
+
+  int result = deflateInit2(zlib_ptr_.get(), comp_level, Z_DEFLATED, window_bits, memory_level,
+                            comp_strategy);
+  is_initialized_ = true;
+
+  return result == Z_OK;
+}
+
+bool ZlibCompressorImpl::init(int window_bits) {
+  // prevents inflateInit2() of being called multiple times, which would cause memory leaks.
+  if (is_initialized_) {
+    return true;
+  }
+  is_deflate_ = false;
+
+  zlib_ptr_->zalloc = Z_NULL;
+  zlib_ptr_->zfree = Z_NULL;
+  zlib_ptr_->opaque = Z_NULL;
+
+  int result = inflateInit2(zlib_ptr_.get(), window_bits);
+  is_initialized_ = true;
+
+  return result == Z_OK;
+}
+
+bool ZlibCompressorImpl::reset() {
+  int result{};
+
+  if (is_deflate_) {
+    result = deflateReset(zlib_ptr_.get());
+  } else {
+    result = inflateReset(zlib_ptr_.get());
+  }
+
+  return result != Z_STREAM_ERROR;
```

Comment:
I'd probably do `return deflateReset(zlib_ptr_.get()) != Z_STREAM_ERROR` and the same for inflateReset which avoids a bit of boilerplate lines. Your call.